### PR TITLE
Add StateWithCovariance and ContactWithCovariance

### PR DESCRIPTION
**Pull request by SyllogismRXS**

### CODE REVIEW

**Comment by SyllogismRXS** [comment]
I originally placed the ContactWithCovariance.h and StateWithCovariance in their respective plugin directories, but after implementing the autonomy side of subscribing to the topics that they publish, I realized that it was tedious to link to the appropriate sensor plugin library. Currently, these two new Covariance data types aren't instantiated in a scrimmage simulation unless a plugin that publishes them, but I think these two data types should become part of scrimmage core.

File: src/plugins/autonomy/MotorSchemas/BehaviorBase.cpp [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+void BehaviorBase::set_max_vector_length(const double &max_vector_length) {
+    max_vector_length_ = max_vector_length;
+}
+
+void BehaviorBase::configure_contacts(std::map<std::string, std::string> &params) {
+    std::vector<std::string> use_contacts;
+    if (get_vec("contacts", params, ", ", use_contacts)) {
+        use_truth_contacts_ = false;
```
[the remark this answers is not in the document]
Comment (SyllogismRXS) [reply src/plugins/autonomy/MotorSchemas/BehaviorBase.cpp]: Sure. it defaults to false, but I can specify it explicitly here too.

File: include/scrimmage/entity/ContactWithCovariance.h [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+ * @section DESCRIPTION
+ * A Long description goes here.
+ *
+ */
+
+#ifndef INCLUDE_SCRIMMAGE_ENTITY_CONTACTWITHCOVARIANCE_H_
+#define INCLUDE_SCRIMMAGE_ENTITY_CONTACTWITHCOVARIANCE_H_
+
+#include <scrimmage/common/ID.h>
+#include <scrimmage/math/StateWithCovariance.h>
```
[the remark this answers is not in the document]
Comment (SyllogismRXS) [reply include/scrimmage/entity/ContactWithCovariance.h]: I was able to rework my commit such that ContactWithCovariance.h doesn't exist. The same Contact.h file is used to contain both State or StateWithCovariance. It's up to the plugin developer to use std::dynamic_pointer_cast to down cast to StateWithCovariance, if needed.

File: include/scrimmage/math/StateWithCovariance.h [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+#ifndef INCLUDE_SCRIMMAGE_MATH_STATEWITHCOVARIANCE_H_
+#define INCLUDE_SCRIMMAGE_MATH_STATEWITHCOVARIANCE_H_
+
+#include <scrimmage/math/State.h>
+
+#include <memory>
+
+namespace scrimmage {
+// x, y, z, rotation about X axis, rotation about Y axis, rotation about Z axis
+using CovarianceMatrix = Eigen::Matrix<double, 6, 6>;
```
[the remark this answers is not in the document]
Comment (SyllogismRXS) [reply include/scrimmage/math/StateWithCovariance.h]: I reworked the StateWithCovariance class such that the plugin developer can choose the Covariance matrix size. There are a number of representations and I don't think we should specify as representation.

File: include/scrimmage/math/StateWithCovariance.h [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+class StateWithCovariance : public scrimmage::State {
+ public:
+    StateWithCovariance();
+    explicit StateWithCovariance(const scrimmage::State &state);
+    explicit StateWithCovariance(const scrimmage::StatePtr &state);
```
[the remark this answers is not in the document]
Comment (SyllogismRXS) [reply include/scrimmage/math/StateWithCovariance.h]: Agreed. Reworked to allow the user to specify covariance matrix size.

File: include/scrimmage/math/StateWithCovariance.h [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+    void set_pose_covariance(const CovarianceMatrix &covariance);
+    void set_velocity_covariance(const CovarianceMatrix &covariance);
+    const CovarianceMatrix & pose_covariance();
+    const CovarianceMatrix & velocity_covariance();
+
+ protected:
+    CovarianceMatrix pose_covariance_;
```
[the remark this answers is not in the document]
Comment (SyllogismRXS) [reply include/scrimmage/math/StateWithCovariance.h]: N/A after cov matrix rework.

File: src/plugins/sensor/NoisyContacts/NoisyContacts.cpp [hunk 6]
```diff
@@ -76,32 +78,39 @@ void NoisyContacts::init(std::map<std::string, std::string> &params) {
...
-    // Create noisy versions of all contacts
+    // Create noisy versions of all contacts that aren't own vehicle
     for (auto &kv: (*parent_->contacts())) {
-        State ns = *(kv.second.state()); // Create copy of contact state
+        // ignore own vehicle
+        if (kv.second.id().id() == parent_->id().id()) {
+            continue;
+        }
+
+        // Construct a new ContactWithCovariance in place
```
[the remark this answers is not in the document]
Comment (SyllogismRXS) [reply src/plugins/sensor/NoisyContacts/NoisyContacts.cpp]: It is specified in the constructor of StateWithCovariance now. The user can also call "set_covariance()" not as well.

**Comment by esquires** [comment]
Thanks. Those changes sound good to me.

**Pull request merged**